### ISSUE TITLE
Recursive errors with BetaModel

### ISSUE THREAD

**Issue by sarangerin** [statsmodels/statsmodels#9048]
#### Describe the bug

[A clear and concise description of what the bug is. This should explain **why** the current behaviour is a problem and why the expected output is a better solution.]

#### Code Sample, a copy-pastable example if possible

```python
from statsmodels.othermod import betareg

beta_glm = betareg.BetaModel(Y_dum_f1.astype(float), X_dum.astype(float))
beta_result = beta_glm.fit()
beta_result.summary()
```
<details>
With same Y and X, I ran a couple of GLM models in statsmodels.api, but with BetaModel I constantly face Error message that is blank:

The expected output is a summary table of the model result. 
My data has a total of 7776 observations, endog is a 1-d array and exog is 7776*10 (10 variables).
<img width="823" alt="Screenshot 2023-10-30 at 12 17 57 PM" src="https://github.com/statsmodels/statsmodels/assets/102279867/9467e0a8-1af7-4b06-b114-3ccd97928784">

**Comment by josef-pkt** [comment]
The failing assert means that your endog y data is not inside the (0, 1) interval.
Note, open interval, 0 and 1 are not included.

This assert could be converted to a proper input check with a intentional raise and an informative error message (similar to Logit, Probit).

What are the values in your `Y_dum_f1.astype(float)` array?

**Comment by josef-pkt** [comment]
I'm not sure I checked 0, 1 observations for Beta model.

AFAIR: for some alpha and Beta, the pdf at 0 and 1 is finite. I'm not sure we can loosen the restriction if we detect those cases. 
In other cases the pdf(0) and pdf(1) go to infinity.

Another option would be to automatically clip the endog to be in the interior of (0, 1) interval. But I don't remember how numerically stable pdf, score and similar are in the 0 and 1 neighborhoods.

**Comment by sarangerin** [comment]
Hi thank you, my engog ranges between 0 and 1, they are probabilities. Now I tried rerunning by deleting the "asssert" part of the code, I see the following result:
<img width="598" alt="Screenshot 2023-10-30 at 12 41 53 PM"

Which I didn't expect as an ideal outcome... src="https://github.com/statsmodels/statsmodels/assets/102279867/feb91f51-c161-4d6f-8bda-d6a60002e48f">

**Comment by josef-pkt** [comment]
score_hessian_factor computes expression like `np.log(y / (1. - y))` which results in +/- inf at the boundaries.
The method includes clipping for expressions for expected params, e.g. alpha and beta (implied by mu and variance), but no clipping for y.

Clipping of y needs to be intentional and not for numerical reasons, i.e. checks and possible workarounds need to be in `model.__init__`. 

Clipping in the parameters is mainly to facilitate optimization. I don't know whether there is a check for the Beta regression similar to perfect separation/prediction in binary models, e.g. warn or raise if alpha or beta go to zero, or predicted mean goes to zero or one for some observations.

**Comment by josef-pkt** [comment]
I guess what should be done here first is to raise an informative Exception if not 0 < y < 1, i.e. in interior of unit interval.

**Comment by ParitoshSinha1990** [comment]
Is it recommended to add a small epsilon to navigate through this error?

**Comment by josef-pkt** [comment]
> Is it recommended to add a small epsilon to navigate through this error?

If you have just a few boundary points, for example because the numbers might have been rounded, then I would add the epsilon shift.

If you have a larger fraction of boundary points, then the Beta model is not the correct model for those data points, e.g. the true model could have zero and/or one inflation. In that case, I would drop the boundary point observations and estimate the beta model only for the interior points.
(We don't have 0-1-inflated Beta model yet.)

One possibility to check the robustness of the estimates, would be to compare the parameter estimates or other results of the model with epsilon corrected observations with a model where boundary observations have been dropped.

If the results differ by a larger amount, then I would investigate why there are boundary points and whether the boundary points really follow the same beta model as the interior points

**Comment by ParitoshSinha1990** [comment]
Thank you for the quick revert. My model base has a large % of observations with value as 0 (essentially click rates). I will try to exclude observations logically (e.g. certain cohorts that are never going to click) before checking basis below

> One possibility to check the robustness of the estimates, would be to compare the parameter estimates or other results of the model with epsilon corrected observations with a model where boundary observations have been dropped.

**Comment by josef-pkt** [comment]
Yes that sounds like zero-inflated. For example there might be a separate logit or binomial model for users that never click versus those that click at least once.

For count data this might be zero-inflated Binomial or similar, but we only have zero-inflated Poisson and Negative Binomial which don't have an upper bound on the count.
